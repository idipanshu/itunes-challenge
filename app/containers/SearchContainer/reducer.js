/*
 *
 * SearchContainer reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: searchContainerTypes, Creators: searchContainerCreators } = createActions({
  requestGetSongs: ['artistName'],
  successGetSongs: ['data'],
  failureGetSongs: ['error'],
  clearSongs: []
});

export const initialState = { artistName: null, songsData: [], songsError: null };

/* eslint-disable default-case, no-param-reassign */
export const searchContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case searchContainerTypes.REQUEST_GET_SONGS:
        draft.artistName = action.artistName;
        break;

      case searchContainerTypes.CLEAR_SONGS:
        draft.artistName = null;
        draft.songsData = [];
        draft.songsError = null;
        break;

      case searchContainerTypes.SUCCESS_GET_SONGS:
        draft.songsData = action.data;
        break;

      case searchContainerTypes.FAILURE_GET_SONGS:
        draft.songsError = get(action.error, 'message', 'something_went_wrong');
        break;
    }
  });

export default searchContainerReducer;
