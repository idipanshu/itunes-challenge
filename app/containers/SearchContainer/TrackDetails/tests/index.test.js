import React from 'react';
import { timeout, renderProvider } from '@utils/testUtils';
import { searchContainerTypes } from '../../reducer';
import { SongDetailsPageTest as SongDetailsPage, mapDispatchToProps } from '../index';

describe('<SongDetailsPage /> tests', () => {
  let submitSpy;
  const params = { trackId: '11111111111' };

  beforeEach(() => {
    submitSpy = jest.fn();
  });

  it('should render and match snapshot', () => {
    const { baseElement } = renderProvider(<SongDetailsPage dispatchGetItunesTracks={submitSpy} match={{ params }} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call dispatchGetItunesTracks when no data is present in the store', async () => {
    const searchedTerm = '11111111111';

    renderProvider(<SongDetailsPage dispatchGetItunesTracks={submitSpy} match={{ params }} />);

    await timeout(500);

    expect(submitSpy).toBeCalledWith(searchedTerm);
  });

  it('should NOT dispatchGetItunesTracks if matching data is found', async () => {
    const trackId = '11111111111';

    renderProvider(
      <SongDetailsPage songsData={[{ trackId }]} dispatchGetItunesTracks={submitSpy} match={{ params }} />
    );

    await timeout(500);

    expect(submitSpy).toBeCalledTimes(0);
  });

  it('should validate mapDispatchToProps actions', async () => {
    const dispatchSongsDataSpy = jest.fn();
    const trackId = '11111111111';

    const actions = {
      dispatchGetItunesTracks: { searchedTerm: trackId, type: searchContainerTypes.REQUEST_GET_SONGS }
    };

    const props = mapDispatchToProps(dispatchSongsDataSpy);
    props.dispatchGetItunesTracks(trackId);
    expect(dispatchSongsDataSpy).toHaveBeenCalledWith(actions.dispatchGetItunesTracks);
  });
});
