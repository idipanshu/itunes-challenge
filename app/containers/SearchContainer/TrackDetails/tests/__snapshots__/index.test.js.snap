// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SongDetailsPage /> tests should render and match snapshot 1`] = `
<body>
  <div>
    <div
      class="TrackDetails__Container-sc-1cwsvw8-0 cleOkp"
    />
  </div>
</body>
`;
